gensim/models/word2vec_inner.pyx

```diff
@@ -343,6 +343,99 @@ def train_sentence_sg(model, sentence, alpha, _work):
     return result
 
 
+def train_batch_sg(model, sentences, alpha, _work):
+    cdef int hs = model.hs
+    cdef int negative = model.negative
+    cdef int sample = (model.sample != 0)
+
+    cdef REAL_t *syn0 = <REAL_t *>(np.PyArray_DATA(model.syn0))
+    cdef REAL_t *word_locks = <REAL_t *>(np.PyArray_DATA(model.syn0_lockf))
+    cdef REAL_t *work
+    cdef REAL_t _alpha = alpha
+    cdef int size = model.layer1_size
+
+    cdef int codelens[MAX_NUM_SENTENCES][MAX_SENTENCE_LEN]
+    cdef np.uint32_t indexes[MAX_NUM_SENTENCES][MAX_SENTENCE_LEN]
+    cdef np.uint32_t reduced_windows[MAX_NUM_SENTENCES][MAX_SENTENCE_LEN]
+    cdef int sentence_len[MAX_NUM_SENTENCES]
+    cdef int window = model.window
+
+    cdef int i, j, k
+    cdef long result = 0
+    cdef int num_sentences = 0
+    cdef int sent_idx = 0
+
+    # For hierarchical softmax
+    cdef REAL_t *syn1
+    cdef np.uint32_t *points[MAX_NUM_SENTENCES][MAX_SENTENCE_LEN]
+    cdef np.uint8_t *codes[MAX_NUM_SENTENCES][MAX_SENTENCE_LEN]
+
+    # For negative sampling
+    cdef REAL_t *syn1neg
+    cdef np.uint32_t *cum_table
+    cdef unsigned long long cum_table_len
+    # for sampling (negative and frequent-word downsampling)
+    cdef unsigned long long next_random
+
+    if hs:
+        syn1 = <REAL_t *>(np.PyArray_DATA(model.syn1))
+
+    if negative:
+        syn1neg = <REAL_t *>(np.PyArray_DATA(model.syn1neg))
+        cum_table = <np.uint32_t *>(np.PyArray_DATA(model.cum_table))
+        cum_table_len = len(model.cum_table)
+    if negative or sample:
+        next_random = (2**24) * model.random.randint(0, 2**24) + model.random.randint(0, 2**24)
+
+    # convert Python structures to primitive types, so we can release the GIL
+    work = <REAL_t *>np.PyArray_DATA(_work)
+
+    vlookup = model.vocab
+    for sent_idx, sentence in enumerate(sentences):
```

Comment:
What happens when `sent_idx` is greater than `MAX_NUM_SENTENCES`?

...as I think happens currently, because it doesn't look like the number of sentences passed from `job_batch` is limited in any way?